distributed/shuffle/_worker_extension.py

```diff
@@ -193,6 +155,359 @@ def heartbeat(self) -> dict[str, Any]:
+    rpc:
+        A callable returning a PooledRPCCall to contact other Shuffle instances.
+        Typically a ConnectionPool.
+    scheduler:
+        A PooledRPCCall to to contact the scheduler.
+    memory_limiter_disk:
+    memory_limiter_comm:
+        A ``ResourceLimiter`` limiting the total amount of memory used in either
+        buffer.
+    """
+
+    def __init__(
+        self,
+        worker_for: dict[NIndex, str],  # TODO
+        output_workers: set,
+        old: ChunkedAxes,
+        new: ChunkedAxes,
+        id: ShuffleId,
+        run_id: int,
+        local_address: str,
+        directory: str,
+        nthreads: int,
+        rpc: Callable[[str], PooledRPCCall],
+        scheduler: PooledRPCCall,
+        memory_limiter_disk: ResourceLimiter,
+        memory_limiter_comms: ResourceLimiter,
+    ):
+        super().__init__(
+            id=id,
+            run_id=run_id,
+            output_workers=output_workers,
+            local_address=local_address,
+            directory=directory,
+            nthreads=nthreads,
+            rpc=rpc,
+            scheduler=scheduler,
+            memory_limiter_comms=memory_limiter_comms,
+            memory_limiter_disk=memory_limiter_disk,
+            dump=dump_shards,
+            load=load_partition,
+        )
+        self.old = old
+        self.new = new
+        partitions_of = defaultdict(list)
+        for part, addr in worker_for.items():
+            partitions_of[addr].append(part)
+        self.partitions_of = dict(partitions_of)
+        self.worker_for = worker_for
+        self._mapping = self._map_old_to_new(old, new)
+        self._old_to_new = _old_to_new(old, new)
+
+    def _map_old_to_new(
+        self, old: ChunkedAxes, new: ChunkedAxes
+    ) -> dict[NIndex, list[tuple[NIndex, NIndex, NSlice]]]:
+
+        # intersections is contains the new individual chunks as combined slices
+        # of the old chunks.
+        #
+        # Each chunk consists of all the n-dimensional slices of the old chunks
+        # that make up the new chunk.
+        #
+        # Each n-dimensional slice contains n tuples consisting of the index of
+        # the old chunk on the n-th dimension and the slice along that dimension.
+        intersections = intersect_chunks(old, new)
+
+        new_indices = product(*(range(len(c)) for c in new))
+
+        ndim = len(old)
+
+        old_to_new = defaultdict(list)
+
+        for new_index, new_chunk in zip(new_indices, intersections):
+            # sub-dimensions of new chunk composed of slices
+            subdims = [
+                len({slice[dim][0] for slice in new_chunk}) for dim in range(ndim)
+            ]
+
+            subdim_indices = product(*(range(dim) for dim in subdims))
+
+            for subdim_index, nchunkslice in zip(subdim_indices, new_chunk):
+                old_index, slices = zip(*nchunkslice)
+                old_to_new[old_index].append((new_index, subdim_index, slices))
+        return old_to_new
```

Comment:
This is actually a static method. Marking it as such or moving it out of the class as a function would allow for straight forward unit testing

Reply:
i.e. I strongly encourage writing a couple of tests for this

Reply:
Done